services/horizon: Move path finding end-points to new actions.

abuiles opened the pull request:
<!-- If you're making a doc PR or something tiny where the below is irrelevant, delete this
template and use a short description, but in your description aim to include both what the
change is, and why it is being made, with enough context for anyone to understand. -->

<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Move path finding handler to /actions and move assertions helper to a share package.

### Why

This is part of https://github.com/stellar/go/issues/2582.

For this handler we need to use `ServerHTTP` since they don't follow the same page building pattern we use for other end-points. Maybe we could wrap it around `restPageHandler` but we'll need to land this first https://github.com/stellar/go/pull/2826/files?file-filters%5B%5D=#diff-b8d16027eadf970091bce20ac2f34450R418 which allow us to specify a page builder. 




### Known limitations

2opremio commented on services/horizon/internal/actions/path.go:
```diff
@@ -116,18 +112,8 @@ var sourceAssetsOrSourceAccount = problem.P{
-	}
 	qp := StrictReceivePathsQuery{}
-	err := actions.GetParams(&qp, r)
+	err := GetParams(&qp, r)
 	if err != nil {
 		problem.Render(ctx, w, err)
 		return
```
```suggestion
	if err := GetParams(&qp, r); err != nil {
		problem.Render(ctx, w, err)
		return
```

abuiles replied:
I dropped this suggestion, it breaks the tests after applying it - we actually can use the `err` variable here since we use it in other places below

2opremio replied:
I don't understand, the code below won't execute of there is an error, so it's equivalent.

2opremio commented on services/horizon/internal/actions/path.go:
```diff
@@ -116,18 +112,8 @@ var sourceAssetsOrSourceAccount = problem.P{
 // ServeHTTP implements the http.Handler interface
 func (handler FindPathsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
Why can't you use `objectActionHandler` or a variant of it?

2opremio replied:
You could use the same for `FeeStatsHandle`

abuiles replied:
For fee stats, we can use objectActionHandler since it returns a JSON object -- unfortunately `objectActionHandler` doesn't fit the bill for this resource not `pageActionHandler` --

The problem with `pageActionHandler` is that it build a a "paginatable" page - but the resource return from path finds doesn't have links, next, or previous.

To fix this nicely, we should allow for a handler which specifies its own page builder -- so we need to solve this https://github.com/stellar/go/pull/2826#discussion_r457581891 and then we can use the solution from there here.

abuiles replied:
@2opremio I added a custom handler for the scenarios where we want to render a  simple base page

abuiles commented:
not the one immediately below, but a few lines below there are other things
that declare `err` too -- this change is fine, but then we have to go and
do something similar in other places (in this same function) where `err`
can happen

On Mon, Jul 20, 2020 at 4:54 PM Alfonso Acosta <notifications@github.com>
wrote:

> *@2opremio* commented on this pull request.
> ------------------------------
>
> In services/horizon/internal/actions/path.go
> <https://github.com/stellar/go/pull/2834#discussion_r457713359>:
>
> > +	err := GetParams(&qp, r)
>  	if err != nil {
>  		problem.Render(ctx, w, err)
>  		return
>
> I don't understand, the code below won't execute of there is an error, so
> it's equivalent.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/stellar/go/pull/2834#discussion_r457713359>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAFKDE43I43USZDWYOXMA3R4S4JPANCNFSM4PCENEUQ>
> .
>

2opremio commented:
> not the one immediately below, but a few lines below there are other things that declare `err` too -- this change is fine, but then we have to go and do something similar in other places (in this same function) where `err` can happen

Sorry, but I still don't get it. What's wrong with re-declaring err below?

abuiles commented:
@2opremio addressed your comments 

About 

> Sorry, but I still don't get it. What's wrong with re-declaring err below?

There is nothing wrong with re-declaring -- I should have explained myself better - is not really that we have to do again `err := fund(); err ` but rather we end up needing a `var err error` -- so at the end I don't think doing the shortes version ends up bringing us any benefit. The code has been updated to use your recommendation.

2opremio commented on services/horizon/internal/handler.go:
```diff
@@ -264,6 +264,47 @@ func (handler objectActionHandler) ServeHTTP(
+
+		httpjson.Render(
+			w,
+			handler.buildPage(response),
```
Can't you use the `BuildPage` interface method from the prior PR?

abuiles replied:
@2opremio no, it  it's a different interface - the previous one is when we want to build a full page, that is a page with  records and paginations  `hal.Page` -- this one is to allow user to build simple pages `hal.BasePage` which only contains a list or resources but no pagination or links data.

2opremio replied:
But `httpjson.Render()` takes an `interface{}` so you could use both (i.e. you could make the`BuildPage` return an `interface{}` instead of `Page` or `BasePage`)

Admittedly, returning an interface is rather ugly though so I will leave it up to you. At the very least, we could avoid having `basePageActionHandler` by reusing `restPageHandler` (by storing an `interface{}`-returning build function)

abuiles replied:
I'm okay having `basePageActionHandler` it makes it clear what type of page the end-point is going to return and we avoid using `interface{}`
